android/src/main/java/io/ably/flutter/plugin/AblyMethodCallHandler.java

```diff
@@ -583,6 +601,132 @@ private void closeRealtime(@NonNull MethodCall call, @NonNull MethodChannel.Resu
+    final AblyFlutterMessage message = (AblyFlutterMessage) call.arguments;
+    this.<AblyFlutterMessage<Map<String, Object>>>ablyDo(message, (ablyLibrary, ablyMessage) -> {
+      final String channelName = (String) ablyMessage.message.get(PlatformConstants.TxTransportKeys.channelName);
+      ablyLibrary.getAblyClient(ablyMessage.handle)
+          .channels
+          .get(channelName)
+          .push.subscribeClientAsync(handleCompletionWithListener(result));
+    });
+  }
+
+  private void pushUnsubscribeClient(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
+    final AblyFlutterMessage message = (AblyFlutterMessage) call.arguments;
+    this.<AblyFlutterMessage<Map<String, Object>>>ablyDo(message, (ablyLibrary, ablyMessage) -> {
+      final String channelName = (String) ablyMessage.message.get(PlatformConstants.TxTransportKeys.channelName);
+      ablyLibrary.getAblyClient(ablyMessage.handle)
+          .channels
+          .get(channelName)
+          .push.unsubscribeClientAsync(handleCompletionWithListener(result));
+    });
+  }
+
+  private void pushListSubscriptions(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
+    final AblyFlutterMessage<?> message = (AblyFlutterMessage<?>) call.arguments;
+    this.<AblyFlutterMessage<Map<String, Object>>>ablyDo(message, (ablyLibrary, ablyMessage) -> {
+      final String channelName = (String) ablyMessage.message.get(PlatformConstants.TxTransportKeys.channelName);
+      final Map<String, Object> paramsMap = (Map<String, Object>) ablyMessage.message.get(PlatformConstants.TxTransportKeys.params);
+
+      ablyLibrary.getAblyClient(ablyMessage.handle)
+          .channels
+          .get(channelName)
+          .push.listSubscriptionsAsync(createParamsArrayFromMap(paramsMap), this.paginatedResponseHandler(result, null));
+    });
+  }
+
+  private Param[] createParamsArrayFromMap(@Nullable Map<String, Object> paramsMap) {
+    if (paramsMap == null) {
+      return new Param[0];
+    }
+    Param[] params = new Param[paramsMap.size()];
+    int paramsSize = 0;
+    for (Map.Entry<String, Object> entry : paramsMap.entrySet()) {
+      Param param = new Param(entry.getKey(), entry.getValue());
+      params[paramsSize] = param;
+      paramsSize += 1;
+    }
+    return params;
+  }
+
+  private void pushDevice(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
+    final AblyFlutterMessage<Integer> message = (AblyFlutterMessage<Integer>) call.arguments;
+    this.<Integer>ablyDo(message, (ablyLibrary, ablyClientHandle) -> {
+      try {
+        AblyRealtime realtime = ablyLibrary.getRealtime(ablyClientHandle);
+        if (realtime != null) {
+          result.success(realtime.device());
+          return;
+        }
+
+        AblyRest rest = ablyLibrary.getRest(ablyClientHandle);
+        if (rest != null) {
+          result.success(rest.device());
+          return;
+        }
```

Comment:
Alternatively we can make use of the newly created `getAblyClient` method. Might save some reading effort?

```java
AblyBase client = ablyLibrary.getAblyClient(ablyClientHandle);
if (client instanceof AblyRealtime) {
  result.success(((AblyRealtime) client).device());
} else if (client!=null) {
  result.success(((AblyRest) client).device());
}
```

Reply:
I wanted to maintain the type safety and I think this is also more readable. Casting and checking `instanceof` is less safe and readable. Additionally, to be safe using the code snippet you provide, we would have to have 3 if/else statements, and in the third one, throw an error. 

This is particularly more confusing because `AblyRealtime` is a `AblyRest` too, so this pattern may be copied in the future, with isinstance AblyRest first and things would break. 

My principle is I only cast as a last resort. WDYT?

Reply:
Makes sense.